pkg/agent/containerd/containerd.go

```diff
@@ -110,15 +112,160 @@ func Run(ctx context.Context, cfg *config.Node) error {
+
+	// At startup all leases from k3s are cleared; we no longer use leases to lock content
+	if err := clearLeases(ctx, client); err != nil {
+		logrus.Errorf("Error while clearing leases: %s", err.Error())
+		return
+	}
+
+	// Clear the pinned labels on all images previously pinned by k3s
+	if err := clearLabels(ctx, client); err != nil {
+		logrus.Errorf("Error while clearing labes: %s", err.Error())
+		return
+	}
+
+	// create the state that watcher will watch
+	stateFileInfos := make(map[string]fs.FileInfo)
+	for _, dirEntry := range fileInfos {
+		if dirEntry.IsDir() {
+			continue
+		}
+
+		// get the file info to add to the state map
+		fileInfo, err := dirEntry.Info()
+		if err != nil {
+			logrus.Errorf("Error while getting the info from file: %s", err.Error())
+			continue
+		}
+
+		// insert the file into the state map that will have the state from the file
+		stateFileInfos[filepath.Join(cfg.Images, dirEntry.Name())] = fileInfo
+	}
+
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				return
+			}
+
+			switch event.Op {
+			case fsnotify.Write:
+				newStateFile, err := os.Stat(event.Name)
+				if err != nil {
+					logrus.Errorf("Error encountered while getting file %s info for event write: %s", event.Name, err.Error())
+					continue
+				}
+
+				// we do not want to handle directorys, only files
+				if newStateFile.IsDir() {
+					continue
+				}
+
+				lastStateFile := stateFileInfos[event.Name]
+				stateFileInfos[event.Name] = newStateFile
+				logrus.Infof("File added to watcher controller: %s", event.Name)
+				if (newStateFile.Size() != lastStateFile.Size()) && newStateFile.ModTime().After(lastStateFile.ModTime()) {
+					logrus.Infof("File met the requirements for import to containerd image store: %s", event.Name)
+					start := time.Now()
+					if err := preloadFile(ctx, cfg, client, imageClient, event.Name); err != nil {
+						logrus.Errorf("Error encountered while importing %s: %v", event.Name, err)
+						continue
+					}
+					logrus.Infof("Imported images from %s in %s", event.Name, time.Since(start))
+				}
+			case fsnotify.Create:
+				info, err := os.Stat(event.Name)
+				if err != nil {
+					logrus.Errorf("Error encountered while getting file %s info for event Create: %s", event.Name, err.Error())
+					continue
+				}
+
+				if info.IsDir() {
+					continue
+				}
+
+				stateFileInfos[event.Name] = info
+				logrus.Infof("File added to watcher controller: %s", event.Name)
+
+				start := time.Now()
+				if err := preloadFile(ctx, cfg, client, imageClient, event.Name); err != nil {
+					logrus.Errorf("Error encountered while importing %s: %v", event.Name, err)
+					continue
+				}
+				logrus.Infof("Imported images from %s in %s", event.Name, time.Since(start))
+			case fsnotify.Rename:
+				delete(stateFileInfos, event.Name)
+				logrus.Infof("Removed file from the watcher controller: %s", event.Name)
+			case fsnotify.Remove:
+				delete(stateFileInfos, event.Name)
```

Comment:
Do we want to un-label the images that came from this file when it is deleted? We'd probably need to use another label to track what file (or files, in case it is in multiple) an image came from to do this.